# Primer Overlaps

This directory contains two scripts that help determine from a bed file the ranges that primers cover.

1. `tb_get_ranges.py` extracts the ranges for each primer pair from a bed file. It assumes that the primers have matching target genes in the names (i.e., atpE-F and atpE-R, etc.) and will create a dictionary of all the ranges for each primer pair, including any primer pairs that target the same gene.
   The input bed file takes the following format:

    - `chromosome	start_position	end_position	primer_name	a number? orientation primer_sequence`; ultimately, though, the only examined columns that matter are `chromosome`, `start_position`, `end_position`, and `primer_name`. 
    - This file produces as output a tab-delimited file with the following format (no header):
        - `primer_name (without any forward/reverse indications)	start_genomic_position	end_genomic_position`

2. `tb_overlapping_ranges.py` will take the output generated by `tb_gene_ranges.py` and determine if multiple primer ranges exist for the same gene. In that case, it will try to combine the ranges if they overlap.
    - The output file takes the following format:
    - `primer_name (without any forward/reverse indications, and any separators (like the -1 and -2 of geneA-1 and geneA-2 removed))	start_genomic_position	end_genomic_position`
  
  This file is used as input to `../coverage_locations/coverage_assessment.py`